packages/docusaurus/src/server/duplicateRoutes.ts

```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+import chalk from 'chalk';
+import {RouteConfig, OnDuplicateRoutes} from '@docusaurus/types';
+import {getFinalRoutes} from './utils';
+import {flatMap} from 'lodash';
+
+export function getAllDuplicateRoutes(
+  pluginsRouteConfigs: RouteConfig[],
+): string[] {
+  const allRoutes: string[] = flatMap(pluginsRouteConfigs, getFinalRoutes).map(
+    (routeConfig) => routeConfig.path,
+  );
+  const seenRoutes: Record<string, any> = {};
+  const duplicateRoutes: string[] = allRoutes.filter(function (route) {
+    if (seenRoutes.hasOwnProperty(route)) {
+      return true;
+    } else {
+      seenRoutes[route] = true;
+      return false;
+    }
+  });
+  return duplicateRoutes;
+}
+
+export function getDuplicateRoutesMessage(
+  allDuplicateRoutes: string[],
+): string {
+  const message = allDuplicateRoutes
+    .map(
+      (duplicateRoute) =>
+        `Attempting to create page at ${duplicateRoute}, but a page already exists at this route`,
+    )
+    .join('\n');
+  return message;
+}
+
+export function handleDuplicateRoutes(
```

Comment:
part of this code is duplicated with broken links, maybe we could make a shared abstraction